tests/nvapi_d3d12.cpp

```diff
@@ -382,6 +392,33 @@ TEST_CASE("D3D12 methods succeed", "[.d3d12]") {
         REQUIRE(commandListRefCount == 0);
     }
 
+    SECTION("Create + Launch CuBIN with SMEM returns OK") {
+        const void* cubinData = nullptr;
+        auto cubinSize = 2U;
+        auto blockX = 3U;
+        auto blockY = 4U;
+        auto blockZ = 5U;
+        auto smemSize = 6U;
+        auto shaderName = "shader";
+        auto shaderHandle = reinterpret_cast<D3D12_CUBIN_DATA_HANDLE*>(0x912122);
+        auto handle = &shaderHandle;
+        const void* params = nullptr;
+        auto paramSize = 7U;
+        const void* rawParam = nullptr;
+        auto rawParamCount = 0U;
+        REQUIRE_CALL(device, CreateCubinComputeShaderWithName(cubinData, cubinSize, blockX, blockY, blockZ, shaderName, handle))
+            .RETURN(S_OK)
+            .TIMES(1);
+        REQUIRE_CALL(commandList, LaunchCubinShaderEx(shaderHandle, blockX, blockY, blockZ, smemSize, params, paramSize, rawParam, rawParamCount))
+            .RETURN(S_OK)
+            .TIMES(1);
+
+        REQUIRE(NvAPI_D3D12_CreateCubinComputeShaderEx(static_cast<ID3D12Device*>(&device), cubinData, cubinSize, blockX, blockY, blockZ, smemSize, shaderName, reinterpret_cast<NVDX_ObjectHandle*>(handle)) == NVAPI_OK);
+        REQUIRE(NvAPI_D3D12_LaunchCubinShader(static_cast<ID3D12GraphicsCommandList*>(&commandList), reinterpret_cast<NVDX_ObjectHandle>(shaderHandle), blockX, blockY, blockZ, params, paramSize) == NVAPI_OK);
+        REQUIRE(deviceRefCount == 0);
+        REQUIRE(commandListRefCount == 0);
+    }
+
```

Comment:
Thanks a lot for the test, I suggest to add another test that validates the behavior when `ID3D12GraphicsCommandListExt1` is not supported, something ilke:

```c++
    SECTION("LaunchCubinShader without ID3D12GraphicsCommandListExt1 returns OK") {
        // Use and configure new commandList to avoid cache hits
        D3D12Vkd3dGraphicsCommandListMock commandList2;
        ALLOW_CALL(commandList2, QueryInterface(__uuidof(ID3D12GraphicsCommandList1), _))
            .LR_SIDE_EFFECT(*_2 = static_cast<ID3D12GraphicsCommandList1*>(&commandList2))
            .LR_SIDE_EFFECT(commandListRefCount++)
            .RETURN(S_OK);
        ALLOW_CALL(commandList2, QueryInterface(ID3D12GraphicsCommandListExt::guid, _))
            .LR_SIDE_EFFECT(*_2 = static_cast<ID3D12GraphicsCommandListExt*>(&commandList2))
            .LR_SIDE_EFFECT(commandListRefCount++)
            .RETURN(S_OK);
        ALLOW_CALL(commandList2, QueryInterface(ID3D12GraphicsCommandListExt1::guid, _))
            .RETURN(E_NOINTERFACE);
        ALLOW_CALL(commandList2, AddRef())
            .LR_SIDE_EFFECT(commandListRefCount++)
            .RETURN(commandListRefCount);
        ALLOW_CALL(commandList2, Release())
            .LR_SIDE_EFFECT(commandListRefCount--)
            .RETURN(commandListRefCount);

        NVDX_ObjectHandle pShader = nullptr;
        auto blockX = 1U;
        auto blockY = 2U;
        auto blockZ = 3U;
        const void* params = nullptr;
        auto paramSize = 4U;
        REQUIRE_CALL(commandList2, LaunchCubinShader(reinterpret_cast<D3D12_CUBIN_DATA_HANDLE*>(pShader), blockX, blockY, blockZ, params, paramSize))
            .RETURN(S_OK)
            .TIMES(1);

        REQUIRE(NvAPI_D3D12_LaunchCubinShader(static_cast<ID3D12GraphicsCommandList*>(&commandList2), pShader, blockX, blockY, blockZ, params, paramSize) == NVAPI_OK);
        REQUIRE(deviceRefCount == 0);
        REQUIRE(commandListRefCount == 0);
    }

```

Reply:
That's a good point, thanks!. I'll add that in with my next batch of updates.

Once I started to grok the mantra behind these tests they made more sense to me :)

Reply:
Cool ;)
(also for my todo list: extract the setup of the mocks at some point in the future to never hit cache issues)